api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java

```diff
@@ -129,4 +155,190 @@ public CohortDefinition getPatientsWhoMissedNextAppointmentAndNotTransferredOut(
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+
+    cd.addSearch(
+        "missedAppointmentLessTransfers",
+        EptsReportUtils.map(
+            getPatientsWhoMissedNextAppointmentAndNotTransferredOut(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "patientsNotFound",
+        EptsReportUtils.map(
+            getPatientsTracedAndNotFound(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "patientsFound",
+        EptsReportUtils.map(
+            getPatientTracedAndFound(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "scheduledLastAppointment",
+        EptsReportUtils.map(
+            getPatientsLastScheduledAppointmentOrDrugPickup(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+
+    cd.setCompositionString(
+        "missedAppointmentLessTransfers AND ((patientsNotFound AND scheduledLastAppointment) AND NOT (patientsFound AND scheduledLastAppointment))");
+
+    return cd;
+  }
+
+  /*
+   a and b and Untraced Patients
+  */
+  public CohortDefinition getPatientsWhoMissedNextAppointmentAndNotTransferredOutAndUntraced() {
+    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+
+    cd.setName("Get patients who missed next appointment, not transferred out and untraced");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+
+    cd.addSearch(
+        "missedAppointmentLessTransfers",
+        EptsReportUtils.map(
+            getPatientsWhoMissedNextAppointmentAndNotTransferredOut(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "withoutVisitCard",
+        EptsReportUtils.map(
+            getAllPatientsWithoutVisitCard(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "patientsWithObs",
+        EptsReportUtils.map(
+            getPatientsWithoutVisitCardAndWithObs(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "scheduledLastAppointment",
+        EptsReportUtils.map(
+            getPatientsLastScheduledAppointmentOrDrugPickup(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+
+    cd.setCompositionString(
+        "missedAppointmentLessTransfers AND ((withoutVisitCard AND scheduledLastAppointment) OR (withoutVisitCard AND scheduledLastAppointment AND patientsWithObs)");
```

Comment:
Does `withoutVisitCard AND scheduledLastAppointment` mean `patients with encounters that are not visit card type and have an appointment scheduled between startDate and endDate`?

I think the previous query was closer to what we want. Maybe we should fetch all patients that do have visit card appointments and exclude them from (a and b).

Reply:
Fixed